File: go/storage.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// ── 数据结构 ──────────────────────────────────────────────────────

type TodoItem struct {
	ID        int    `json:"id"`
	Title     string `json:"title"`
	Completed bool   `json:"completed"`
	CreatedAt string `json:"created_at"`
	UpdatedAt string `json:"updated_at"`
}

type DataFile struct {
	Todos  []TodoItem `json:"todos"`
	NextID int        `json:"next_id"`
}

// ── 存储层 ────────────────────────────────────────────────────────

type TodoStorage struct {
	path   string
	items  []TodoItem
	nextID int
}

func NewTodoStorage(path string) *TodoStorage {
	s := &TodoStorage{
		path:   path,
		items:  make([]TodoItem, 0),
		nextID: 1,
	}
	s.load()
	return s
}

// ---- 文件 I/O ----

func (s *TodoStorage) load() {
	data, err := os.ReadFile(s.path)
	if err != nil {
		// 文件不存在 → 空列表
		return
	}
	var file DataFile
	if err := json.Unmarshal(data, &file); err != nil {
		fmt.Fprintf(os.Stderr, "错误：JSON 解析失败 - %v\n", err)
		os.Exit(1)
	}
	s.items = file.Todos
	s.nextID = file.NextID
}

func (s *TodoStorage) save() {
	// 确保目录存在
	dir := filepath.Dir(s.path)
	os.MkdirAll(dir, 0755)

	file := DataFile{
		Todos:  s.items,
		NextID: s.nextID,
	}
	data, err := json.MarshalIndent(file, "", "  ")
	if err != nil {
		fmt.Fprintf(os.Stderr, "错误：JSON 序列化失败 - %v\n", err)
		os.Exit(1)
	}
	if err := os.WriteFile(s.path, data, 0644); err != nil {
		fmt.Fprintf(os.Stderr, "错误：无法写入 %s - %v\n", s.path, err)
		os.Exit(1)
	}
}

func now() string {
	return time.Now().UTC().Format("2006-01-02T15:04:05Z")
}

func fmtDisplayTime(iso string) string {
	// "2025-01-01T10:00:00Z" → "2025-01-01 10:00:00 +0000"
	if len(iso) < 20 {
		return iso
	}
	s := strings.Replace(iso, "T", " ", 1)
	if strings.HasSuffix(s, "Z") {
		s = s[:len(s)-1] + " +0000"
	}
	return s
}

// ---- 命令实现 ----

func (s *TodoStorage) Add(title string) error {
	if title == "" {
		return fmt.Errorf("标题不能为空")
	}
	nowStr := now()
	item := TodoItem{
		ID:        s.nextID,
		Title:     title,
		Completed: false,
		CreatedAt: nowStr,
		UpdatedAt: nowStr,
	}
	s.nextID++
	fmt.Printf("已添加：ID %d - %s\n", item.ID, item.Title)
	s.items = append(s.items, item)
	s.save()
	return nil
}

func (s *TodoStorage) List() error {
	if len(s.items) == 0 {
		fmt.Println("暂无任务")
		return nil
	}

	// 按 id 升序
	sorted := make([]TodoItem, len(s.items))
	copy(sorted, s.items)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].ID < sorted[j].ID })

	fmt.Println("ID  状态  标题                创建时间")
	fmt.Println("─── ──── ──────────────────── ─────────────────────────")

	for _, item := range sorted {
		status := "☐"
		if item.Completed {
			status = "☑"
		}
		fmt.Printf("%3d  %-4s %-20s  %s\n",
			item.ID, status, item.Title, fmtDisplayTime(item.CreatedAt))
	}
	return nil
}

func (s *TodoStorage) Done(id int) error {
	for i := range s.items {
		if s.items[i].ID == id {
			if s.items[i].Completed {
				fmt.Printf("任务 ID %d 已经是完成状态\n", id)
				return nil
			}
			s.items[i].Completed = true
			s.items[i].UpdatedAt = now()
			fmt.Printf("已完成：ID %d - %s\n", s.items[i].ID, s.items[i].Title)
			s.save()
			return nil
		}
	}
	return fmt.Errorf("ID %d 不存在", id)
}

func (s *TodoStorage) Delete(id int) error {
	for i, item := range s.items {
		if item.ID == id {
			s.items = append(s.items[:i], s.items[i+1:]...)
			fmt.Printf("已删除：ID %d - %s\n", item.ID, item.Title)
			s.save()
			return nil
		}
	}
	return fmt.Errorf("ID %d 不存在", id)
}

func (s *TodoStorage) Edit(id int, newTitle string) error {
	if newTitle == "" {
		return fmt.Errorf("标题不能为空")
	}
	for i := range s.items {
		if s.items[i].ID == id {
			s.items[i].Title = newTitle
			s.items[i].UpdatedAt = now()
			fmt.Printf("已更新：ID %d - %s\n", s.items[i].ID, s.items[i].Title)
			s.save()
			return nil
		}
	}
	return fmt.Errorf("ID %d 不存在", id)
}

func (s *TodoStorage) ClearDone() error {
	before := len(s.items)
	kept := make([]TodoItem, 0, len(s.items))
	for _, item := range s.items {
		if !item.Completed {
			kept = append(kept, item)
		}
	}
	removed := before - len(kept)
	if removed == 0 {
		fmt.Println("没有已完成的任务需要清除")
		return nil
	}
	s.items = kept
	fmt.Printf("已清除 %d 条已完成任务\n", removed)
	s.save()
	return nil
}

func Help() {
	fmt.Println("用法：todo [--file <路径>] <命令> [参数...]")
	fmt.Println()
	fmt.Println("命令：")
	fmt.Println("  add <标题>        添加任务")
	fmt.Println("  list              列出所有任务")
	fmt.Println("  done <ID>         标记任务为已完成")
	fmt.Println("  delete <ID>       删除任务")
	fmt.Println("  edit <ID> <标题>  修改任务标题")
	fmt.Println("  clear             清除所有已完成任务")
	fmt.Println("  help              显示此帮助")
	fmt.Println()
	fmt.Println("全局选项：")
	fmt.Println("  --file <路径>     指定数据文件路径（默认：~/.todo/todos.json）")
}

File: rust/src/storage.rs
use chrono::Utc;
use serde::{Deserialize, Serialize};
use std::fs;
use std::io::{BufReader, BufWriter};
use std::path::PathBuf;

// ── 数据结构 ──────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TodoItem {
    pub id: u32,
    pub title: String,
    pub completed: bool,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct DataFile {
    todos: Vec<TodoItem>,
    next_id: u32,
}

// ── 存储层 ────────────────────────────────────────────────────────

pub struct TodoStorage {
    path: PathBuf,
    items: Vec<TodoItem>,
    next_id: u32,
}

impl TodoStorage {
    pub fn new(path: PathBuf) -> Self {
        let mut s = Self {
            path,
            items: Vec::new(),
            next_id: 1,
        };
        s.load();
        s
    }

    // ---- 文件 I/O ----

    fn load(&mut self) {
        let file = match fs::File::open(&self.path) {
            Ok(f) => f,
            Err(_) => return, // 文件不存在 → 空列表
        };
        let reader = BufReader::new(file);
        let data: DataFile = match serde_json::from_reader(reader) {
            Ok(d) => d,
            Err(e) => {
                eprintln!("错误：JSON 解析失败 - {}", e);
                std::process::exit(1);
            }
        };
        self.items = data.todos;
        self.next_id = data.next_id;
    }

    fn save(&self) {
        // 确保目录存在
        if let Some(parent) = self.path.parent() {
            fs::create_dir_all(parent).ok();
        }

        let data = DataFile {
            todos: self.items.clone(),
            next_id: self.next_id,
        };
        let file = fs::File::create(&self.path).unwrap_or_else(|e| {
            eprintln!("错误：无法写入 {} - {}", self.path.display(), e);
            std::process::exit(1);
        });
        let writer = BufWriter::new(file);
        serde_json::to_writer_pretty(writer, &data).unwrap_or_else(|e| {
            eprintln!("错误：JSON 写入失败 - {}", e);
            std::process::exit(1);
        });
    }

    fn now() -> String {
        Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string()
    }

    fn fmt_display_time(iso: &str) -> String {
        // "2025-01-01T10:00:00Z" → "2025-01-01 10:00:00 +0000"
        if iso.len() < 20 {
            return iso.to_string();
        }
        let mut s = iso.to_string();
        s.replace_range(10..11, " "); // T → 空格
        if s.ends_with('Z') {
            s.truncate(s.len() - 1);
            s.push_str(" +0000");
        }
        s
    }

    // ---- 命令实现 ----

    pub fn add(&mut self, title: &str) -> i32 {
        if title.is_empty() {
            eprintln!("错误：标题不能为空");
            std::process::exit(1);
        }
        let now = Self::now();
        let item = TodoItem {
            id: self.next_id,
            title: title.to_string(),
            completed: false,
            created_at: now.clone(),
            updated_at: now,
        };
        self.next_id += 1;
        println!("已添加：ID {} - {}", item.id, item.title);
        self.items.push(item);
        self.save();
        0
    }

    pub fn list(&self) -> i32 {
        if self.items.is_empty() {
            println!("暂无任务");
            return 0;
        }

        let mut sorted = self.items.clone();
        sorted.sort_by_key(|i| i.id);

        println!("ID  状态  标题                创建时间");
        println!("─── ──── ──────────────────── ─────────────────────────");

        for item in &sorted {
            let status = if item.completed { "☑" } else { "☐" };
            println!(
                "{:>3}  {:<4} {:<20}  {}",
                item.id,
                status,
                item.title,
                Self::fmt_display_time(&item.created_at)
            );
        }
        0
    }

    pub fn done(&mut self, id: u32) -> i32 {
        let item = match self.items.iter_mut().find(|i| i.id == id) {
            Some(i) => i,
            None => {
                eprintln!("错误：ID {} 不存在", id);
                return 1;
            }
        };
        if item.completed {
            println!("任务 ID {} 已经是完成状态", id);
            return 0;
        }
        item.completed = true;
        item.updated_at = Self::now();
        println!("已完成：ID {} - {}", item.id, item.title);
        self.save();
        0
    }

    pub fn delete(&mut self, id: u32) -> i32 {
        let pos = match self.items.iter().position(|i| i.id == id) {
            Some(p) => p,
            None => {
                eprintln!("错误：ID {} 不存在", id);
                return 1;
            }
        };
        let item = self.items.remove(pos);
        println!("已删除：ID {} - {}", item.id, item.title);
        self.save();
        0
    }

    pub fn edit(&mut self, id: u32, new_title: &str) -> i32 {
        if new_title.is_empty() {
            eprintln!("错误：标题不能为空");
            std::process::exit(1);
        }
        let item = match self.items.iter_mut().find(|i| i.id == id) {
            Some(i) => i,
            None => {
                eprintln!("错误：ID {} 不存在", id);
                return 1;
            }
        };
        item.title = new_title.to_string();
        item.updated_at = Self::now();
        println!("已更新：ID {} - {}", item.id, item.title);
        self.save();
        0
    }

    pub fn clear_done(&mut self) -> i32 {
        let before = self.items.len();
        self.items.retain(|i| !i.completed);
        let removed = before - self.items.len();
        if removed == 0 {
            println!("没有已完成的任务需要清除");
            return 0;
        }
        println!("已清除 {} 条已完成任务", removed);
        self.save();
        0
    }

    pub fn help() -> i32 {
        println!("用法：todo [--file <路径>] <命令> [参数...]");
        println!();
        println!("命令：");
        println!("  add <标题>        添加任务");
        println!("  list              列出所有任务");
        println!("  done <ID>         标记任务为已完成");
        println!("  delete <ID>       删除任务");
        println!("  edit <ID> <标题>  修改任务标题");
        println!("  clear             清除所有已完成任务");
        println!("  help              显示此帮助");
        println!();
        println!("全局选项：");
        println!("  --file <路径>     指定数据文件路径（默认：~/.todo/todos.json）");
        0
    }
}

// ── 测试 ──────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    fn temp_path(name: &str) -> PathBuf {
        let mut p = std::env::temp_dir();
        p.push(format!(
            "todo_test_{}_{}.json",
            std::process::id(),
            name
        ));
        // 如果已存在，先删除
        let _ = fs::remove_file(&p);
        p
    }

    fn with_storage<F>(name: &str, f: F)
    where
        F: FnOnce(&mut TodoStorage),
    {
        let path = temp_path(name);
        let mut s = TodoStorage::new(path.clone());
        f(&mut s);
        let _ = fs::remove_file(&path);
    }

    #[test]
    fn test_add_todo() {
        with_storage("add_todo", |s| {
            assert_eq!(s.add("测试任务1"), 0);
            assert_eq!(s.items.len(), 1);
            assert_eq!(s.items[0].title, "测试任务1");
        });
    }

    #[test]
    fn test_list_empty() {
        let path = temp_path("list_empty");
        let s = TodoStorage::new(path.clone());
        // list() 只是打印，不 panic 即可
        assert_eq!(s.list(), 0);
        let _ = fs::remove_file(&path);
    }

    #[test]
    fn test_done() {
        with_storage("done", |s| {
            s.add("可完成");
            assert_eq!(s.done(1), 0);
            assert!(s.items[0].completed);
        });
    }

    #[test]
    fn test_done_already_completed() {
        with_storage("done_already", |s| {
            s.add("已完成");
            s.done(1);
            assert_eq!(s.done(1), 0); // 幂等，不报错
            assert!(s.items[0].completed);
        });
    }

    #[test]
    fn test_delete() {
        with_storage("delete", |s| {
            s.add("待删除");
            assert_eq!(s.delete(1), 0);
            assert!(s.items.is_empty());
        });
    }

    #[test]
    fn test_edit() {
        with_storage("edit", |s| {
            s.add("旧标题");
            assert_eq!(s.edit(1, "新标题"), 0);
            assert_eq!(s.items[0].title, "新标题");
        });
    }

    #[test]
    fn test_id_not_found() {
        with_storage("id_not_found", |s| {
            assert_eq!(s.done(999), 1);
            assert_eq!(s.delete(999), 1);
            assert_eq!(s.edit(999, "任意"), 1);
        });
    }

    #[test]
    fn test_clear_done() {
        with_storage("clear_done", |s| {
            s.add("任务 A");
            s.add("任务 B");
            s.done(2);
            assert_eq!(s.clear_done(), 0);
            assert_eq!(s.items.len(), 1);
            assert_eq!(s.items[0].id, 1);
        });
    }

    #[test]
    fn test_clear_idempotent() {
        with_storage("clear_idempotent", |s| {
            s.add("只有未完成");
            assert_eq!(s.clear_done(), 0);
            assert_eq!(s.items.len(), 1);
        });
    }

    #[test]
    fn test_persistence() {
        let path = temp_path("persistence");
        {
            let mut s = TodoStorage::new(path.clone());
            s.add("持久化测试");
        }
        {
            let s = TodoStorage::new(path.clone());
            assert_eq!(s.items.len(), 1);
            assert_eq!(s.items[0].title, "持久化测试");
        }
        let _ = fs::remove_file(&path);
    }

    #[test]
    fn test_file_not_exists() {
        let path = PathBuf::from("/tmp/todo_test_nonexistent_ok.json");
        let s = TodoStorage::new(path.clone());
        assert_eq!(s.list(), 0);
        let _ = fs::remove_file(&path);
    }
}
